File: collatz/collatz_iterative.py
# Justin Caringal
# an iterative implementation of the collatz conjecture

DESCRIPTION = '''\t\t---Collatz Conjecture---

\t Program by J. Caringal

Under the rules
\t-If the number is even, divide it by two
\t-If the number is odd, triple it and add one
will every positive integer eventually transform into 1?

Iterative Python implementation.

'''

def get_user_input():
    """Gets user input
    
    A function to organize user input and validation

    Args:
        None

    Returns:
        int: Returns the number to be used in the conjecture
    """

    # bool flag to ask for inputs
    is_int = False

    while not is_int:
        # prompts user for integer
        input_num = input('Give me an integer: ')

        # converts str input into int
        try:
            manipulated_num = int(input_num)
        except:
            print('Input not accepted, try again.')
            continue

        is_int = True # is ignored if exception handling sends loop to top

    return manipulated_num


def collatz_iterative(manipulated_num):
    """Implements collatz
    
    A function that implements an iterative implementation
    of the collatz conjecture
    
    Args:
        manipulated_num (int): A user-defined number
        
    Returns:
        int: returns the number of operations performed
    """

    # implements iteration counter (the number of
    # operations performed before reaching 1)
    counter = 0

    # main driving loop
    while manipulated_num != 1:
        
        # increments counter
        counter += 1

        if manipulated_num % 2 == 0: # even
            manipulated_num /= 2
        else: # odd
            manipulated_num = (3 * manipulated_num) + 1
        
        # truncates posibility of floats into an int
        manipulated_num = int(manipulated_num)

        print(f'{counter : <5}{manipulated_num}')
    
    return counter


def main():
    """ THE MAIN FUNCTION """

    print(DESCRIPTION)

    # input num
    manipulated_num = get_user_input()

    # main driver function for collatz
    counter = collatz_iterative(manipulated_num)

    if counter == 1:
        print('It took 1 iteration to reach 1.')
    else:
        print(f'It took {counter} iterations to reach 1.')
    
    print('Exiting program.')


if __name__ == '__main__':
    main()

File: collatz/collatz_iterative.cpp
// Justin Caringal
// an iterative implementation of the collatz conjecture

#include <iostream>
#include <string>

using namespace std;

const string DESCRIPTION = {"\t\t---Collatz Conjecture---\n\n"
    "\t Program by J. Caringal\n\nUnder the rules"
    "\n\t-If the number is even, divide it by two"
    "\n\t-If the number is odd, triple it and add one"
    "\nwill every positive integer eventually transform into 1?\n\n"
    "Iterative C++ implementation.\n\n"};


/**
 * @brief checks to see if inputted string is a valid integer
 * 
 * @param input the string to be tested
 * @return true if input is valid integer
 * @return false if input is not an integer
 */
bool check_int(string input);

/**
 * @brief Get the user input and validates input
 * 
 * @return int : the user-defined value that will be used for the conjecture
 */
int get_user_input();


int main()
{
    cout << DESCRIPTION;
    int test = get_user_input();
}


bool check_int(string input)
{
    // iterates through entire string
    for (size_t index = 0; index < input.length(); index++)
    {
        // checks each character to see if it's valid
        if (isdigit(input[index]) == false)
        {
            return false;
        }
    }

    // all of the chars in the string are digits,
    // therefore it's a valid int
    return true;
}


int get_user_input()
{
    // declares input variable, initializes with placeholder
    string input_num = "DUMMY";

    // main driving loop for user input
    while (!check_int(input_num))
    {
        // prompts user to give input
        cout << "Give me an integer: ";
        cin >> input_num;
    }

    int cleaned_input = stoi(input_num);
    cout << cleaned_input << " success!";
    return cleaned_input;
}